### PR TITLE
Doc: Document difference between agents

### PR DESCRIPTION
**Pull request by JayQuimby**
## Issue #628 - Document difference between different agents (and maybe expose descriptions in UI)

- Updated Documentation in all `Agenthub/...` files to docstrings
- Small prompt change in `codeact_agent` for consistency
- Bug fix in Monologue: LLM needed to be imported
- Renamed `docs/examples` to `docs/documentation`

Feel free to delete `docs/documentation/Agents.md` if you feel it is unnecessary. I thought it would be nice just to compare them to each other, but up to you.

### CODE REVIEW

File: agenthub/codeact_agent/codeact_agent.py [hunk 1]
```diff
@@ -61,13 +66,27 @@ def __init__(
...
+        """
+        Performs one step using the Code Act Agent. 
+        This includes gathering info on previous steps and prompting the model to make a command to execute.
+
+        Parameters:
+        - state (State): used to get updated info and background commands
+
+        Returns:
+        - CmdRunAction(command) - command action to run
+        - AgentEchoAction(content=INVALID_INPUT_MESSAGE) - invalid command ouput
```
Comment (enyst) [inline agenthub/codeact_agent/codeact_agent.py:84]: s/ouput/output

File: agenthub/monologue_agent/agent.py [hunk 2]
```diff
@@ -77,14 +77,34 @@
...
+    Responsible for talking to the user and acting as an overall project manager. 
+    The monologue agent has a long term memory that it is capable of modifying to suit its current needs. 
+    This agent is mostly responsible for managing state and long term memory.
+    """
+
     _initialized = False
 
     def __init__(self, llm: LLM):
+        """
+        Initilizes the Monologue Agent with an llm, monologue, and memory.
```
Comment (enyst) [inline agenthub/monologue_agent/agent.py:90]: s/Initilizes/Initializes :)

File: agenthub/monologue_agent/agent.py [hunk 3]
```diff
@@ -95,7 +115,18 @@ def _add_event(self, event: dict):
         if self.monologue.get_total_length() > MAX_MONOLOGUE_LENGTH:
             self.monologue.condense(self.llm)
 
-    def _initialize(self, task):
+    def _initialize(self, task: str):
+        """
+        Utilizes the INTIAL_THOUGHTS list to give the agent a context for it's capabilities and how to navigate the /workspace.
+        Short circuted to return when already initilized.
```
Comment (enyst) [inline agenthub/monologue_agent/agent.py:121]: INTIAL_THOUGHTS -> INITIAL_THOUGHTS
it's capabilities -> its capabilities
initilized -> initialized

File: agenthub/monologue_agent/agent.py [hunk 4]
```diff
@@ -95,7 +115,18 @@ def _add_event(self, event: dict):
...
+
+        Parameters:
+        - task (str): The initial goal statment provided by the user
```
Comment (enyst) [inline agenthub/monologue_agent/agent.py:124]: s/statment/statement

File: agenthub/monologue_agent/agent.py [hunk 5]
```diff
@@ -148,6 +179,15 @@ def _initialize(self, task):
         self._initialized = True
 
     def step(self, state: State) -> Action:
+        """
+        Modifies the current state by adding the most rescent actions and observations, then prompts the model to think about its next action to take.
```
Comment (enyst) [inline agenthub/monologue_agent/agent.py:183]: s/rescent/recent

File: agenthub/monologue_agent/agent.py [hunk 6]
```diff
@@ -168,5 +208,14 @@ def step(self, state: State) -> Action:
         return action
 
     def search_memory(self, query: str) -> List[str]:
+        """
+        Uses VectorIndexRetriever to find related memories withing the long term memory.
```
Comment (enyst) [inline agenthub/monologue_agent/agent.py]: s/withing/within

File: agenthub/monologue_agent/utils/json.py [hunk 7]
```diff
@@ -2,13 +2,29 @@
...
 
 def my_encoder(obj):
+    """
+    Encodes objects as dictionaries
+
+    Parameters:
+    - obj (Object): An object that will be converted
+
+    Returns:
+    - dict: It the object can be converted it is returned in dict format
```
Comment (enyst) [inline agenthub/monologue_agent/utils/json.py]: s / It the object / If the object

File: agenthub/monologue_agent/utils/json.py [hunk 8]
```diff
@@ -2,13 +2,29 @@
...
+    """
     if hasattr(obj, "to_dict"):
         return obj.to_dict()
 
 def dumps(obj, **kwargs):
+    """
+    Serialize and object to str format
```
Comment (enyst) [inline agenthub/monologue_agent/utils/json.py:19]: Serialize an object

File: agenthub/monologue_agent/utils/memory.py [hunk 9]
```diff
@@ -40,14 +40,28 @@
...
 
 class LongTermMemory:
+    """
+    Responsible for storing information that the agent can call on later for better insights and context.
+    Uses chromadb to store and search through memories.
+    """
+
     def __init__(self):
+        """
+        Initilize the chromadb and set up ChromaVectorStore for later use.
```
Comment (enyst) [inline agenthub/monologue_agent/utils/memory.py:50]: Initialize

File: agenthub/monologue_agent/agent.py [hunk 10]
```diff
@@ -168,5 +208,14 @@ def step(self, state: State) -> Action:
...
 
     def search_memory(self, query: str) -> List[str]:
+        """
+        Uses VectorIndexRetriever to find related memories within the long term memory.
+
+        Parameters:
+        - query (str): The query that we want to find related memories for
+
+        Returns:
+        - List[str]: A list of all text results that matched the query
```
Comment (enyst) [inline agenthub/monologue_agent/agent.py:218]: Isn't this a list of top 10 text results?

File: agenthub/monologue_agent/utils/monologue.py [hunk 11]
```diff
@@ -1,21 +1,50 @@
...
 
 class Monologue:
+    """
+    The monologue is a representation for the agent's internal monologue where it can think.
+    The agent has the capability of using this monologue for whatever it wants.
+    """
+
     def __init__(self):
+        """
+        Initilize the empty list of thoughts
```
Comment (enyst) [inline agenthub/monologue_agent/utils/monologue.py]: Initialize :)

File: agenthub/monologue_agent/utils/monologue.py [hunk 12]
```diff
@@ -1,21 +1,50 @@
...
+        Raises:
+        - ValueError: If t is not a dict
+        """
         if not isinstance(t, dict):
             raise ValueError("Event must be a dictionary")
         self.thoughts.append(t)
 
     def get_thoughts(self):
+        """
+        Get the current toughts of the agent.
```
Comment (enyst) [inline agenthub/monologue_agent/utils/monologue.py]: s/toughts/thoughts

File: agenthub/monologue_agent/utils/prompts.py [hunk 13]
```diff
@@ -87,7 +87,13 @@
 """
 
 
-def get_summarize_monologue_prompt(thoughts):
+def get_summarize_monologue_prompt(thoughts: List[dict]):
+    """
+    Gets the prompt for summarizing the monologue
+
+    Retruns:
```
Comment (enyst) [inline agenthub/monologue_agent/utils/prompts.py:94]: Returns

File: agenthub/monologue_agent/utils/prompts.py [hunk 14]
```diff
@@ -97,6 +103,18 @@ def get_request_action_prompt(
         thoughts: List[dict],
         background_commands_obs: List[CmdOutputObservation] = [],
 ):
+    """
+    Gets the action prompt formatted with appropriate values
+
+    Paramters:
+    - task (str): The current task the agent is trying to accomplish
+    - thoughts (List[dict]): The agents current toughts
```
Comment (enyst) [inline agenthub/monologue_agent/utils/prompts.py:111]: s/ agents current toughts / agent's current thoughts :)

File: agenthub/monologue_agent/utils/prompts.py [hunk 15]
```diff
@@ -97,6 +103,18 @@ def get_request_action_prompt(
         thoughts: List[dict],
         background_commands_obs: List[CmdOutputObservation] = [],
 ):
+    """
+    Gets the action prompt formatted with appropriate values
+
+    Paramters:
```
Comment (enyst) [inline agenthub/monologue_agent/utils/prompts.py]: Parameters

File: agenthub/planner_agent/agent.py [hunk 16]
```diff
@@ -8,10 +8,33 @@
 from opendevin.action import Action
 
 class PlannerAgent(Agent):
+    """
+    The planner agent is responsible for looking at the current progress of the task as well as the goal and evaluating the best coarse of action given the state.
+    This agent will respond with either a though or an action.
```
Comment (enyst) [inline agenthub/planner_agent/agent.py]: s/ coarse / course
s/ though / thought

File: agenthub/planner_agent/agent.py [hunk 17]
```diff
@@ -8,10 +8,33 @@
...
+    """
+
     def __init__(self, llm: LLM):
+        """
+        Initilize the Planner Agent with an LLM
```
Comment (enyst) [inline agenthub/planner_agent/agent.py:18]: Initialize

**Comment by JayQuimby** [comment]
@enyst Thank you so much!

File: agenthub/planner_agent/prompt.py [hunk 18]
```diff
@@ -129,7 +129,18 @@
 %(hint)s
 """
 
-def get_prompt(plan: Plan, history: List[Tuple[Action, Observation]]):
+def get_prompt(plan: Plan, history: List[Tuple[Action, Observation]]) -> str:
+    """
+    Returns the prompt for the planner agent, formatted with the most rescent actions and observations in history
```
Comment (enyst) [inline agenthub/planner_agent/prompt.py:134]: s/rescent/recent

File: docs/documentation/Agents.md [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+`Observation`,
+`NullObservation`,
+`CmdOutputObservation`,
+`FileReadObservation`,
+`AgentRecallObservation`,
+`BrowserOutputObservation`
+
+
+### Methods:
+`__init__`: Initilizes the agent with a long term memory, and an internal monologue
```
Comment (enyst) [inline docs/documentation/Agents.md:28]: Initializes :)

File: docs/documentation/Agents.md [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+
+`_add_event`: Appends events to the monologue of the agent and condenses with summary automatically if the monlogue is too long
```
Comment (enyst) [inline docs/documentation/Agents.md:30]: monologue

File: docs/documentation/Agents.md [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+`_initilize`: Utilizes the `INTIAL_THOUGHTS` list to give the agent a context for it's capabilities and how to navigate the `/workspace`
```
Comment (enyst) [inline docs/documentation/Agents.md:32]: initialize ... INITIAL ... for its capabilities

File: docs/documentation/Agents.md [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+`step`: Modifies the current state by adding the most rescent actions and observations, then prompts the model to think about its next action to take. 
+
+`search_memeory`: Uses `VectorIndexRetriever` to find related memories withing the long term memory.
```
Comment (enyst) [inline docs/documentation/Agents.md:36]: s/withing/within

File: docs/documentation/Agents.md [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+
+## Planner Agent:
+
+### Description:
+The planner agent is responsible for looking at the current progress of the task as well as the goal and evaluating the best coarse of action given the state. This agent will respond with either a though or an action.
```
Comment (enyst) [inline docs/documentation/Agents.md:41]: coarse > course
though > thought

File: docs/documentation/Agents.md [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+### Observations:
+`Observation`,
+`NullObservation`,
+`CmdOutputObservation`,
+`FileReadObservation`,
+`AgentRecallObservation`,
+`BrowserOutputObservation`
+
+### Methods:
+`__init__`: Initilizes an agent with `llm`
```
Comment (enyst) [inline docs/documentation/Agents.md:67]: Initializes

File: docs/documentation/Agents.md (hunk 22 above)
Comment (enyst) [inline docs/documentation/Agents.md:36]: search_memory

File: docs/documentation/Agents.md [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+`CmdRunAction`,
+`AgentEchoAction`,
+`AgentFinishAction`,
+
+### Observations:
+`CmdOutputObservation`,
+`AgentMessageObservation`,
+
+### Methods:
+`__init__`: Initilizes an agent with `llm` and a list of messages `List[Mapping[str, str]]`
```
Comment (enyst) [inline docs/documentation/Agents.md:89]: Initializes

**Comment by enyst** [comment]
Thank you, this is awesome!

**Comment by rbren** [comment]
Thanks for doing this!

There's one bit of confusion that I think would be helpful to clear up: the current agents aren't really specialists in any way. They're meant to be standalone agents--they all do all the things. Here's how I'd describe each one:
* Codeact: a minimalist agent
* Monologue: an agent with long-term and short-term memory management
* Planner: an agent that utilizes the Plan mechanism to plot a long-term course of action

**Comment by JayQuimby** [comment]
> the current agents aren't really specialists in any way. They're meant to be standalone agents

This makes a lot of sense now that you said it. I can go rewrite the confusing parts. Thanks for that clarification.

**Comment by JayQuimby** [comment]
@rbren I have been looking at Monologue Agent, and it looks like there is no utilization of the `LongTermMemory`. Is this intentional? Specifically, I can't find any instances of `MonologueAgent.search_for()` which would use the long term memory.  Should I make an issue for this?

**Review by rbren — APPROVED** [review]

**Pull request merged**